cpp/ql/src/Security/CWE/CWE-319/UseOfHttp.ql

```diff
@@ -0,0 +1,78 @@
+/**
+ * @name Failure to use HTTPS URLs
+ * @description Non-HTTPS connections can be intercepted by third parties.
+ * @kind path-problem
+ * @problem.severity warning
+ * @precision medium
+ * @id cpp/non-https-url
+ * @tags security
+ *       external/cwe/cwe-319
+ */
+
+import cpp
+import semmle.code.cpp.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * A string matching private host names of IPv4 and IPv6, which only matches
+ * the host portion therefore checking for port is not necessary.
+ * Several examples are localhost, reserved IPv4 IP addresses including
+ * 127.0.0.1, 10.x.x.x, 172.16.x,x, 192.168.x,x, and reserved IPv6 addresses
+ * including [0:0:0:0:0:0:0:1] and [::1]
+ */
+class PrivateHostName extends string {
+  bindingset[this]
+  PrivateHostName() {
+    this.regexpMatch("(?i)localhost(?:[:/?#].*)?|127\\.0\\.0\\.1(?:[:/?#].*)?|10(?:\\.[0-9]+){3}(?:[:/?#].*)?|172\\.16(?:\\.[0-9]+){2}(?:[:/?#].*)?|192.168(?:\\.[0-9]+){2}(?:[:/?#].*)?|\\[?0:0:0:0:0:0:0:1\\]?(?:[:/?#].*)?|\\[?::1\\]?(?:[:/?#].*)?")
+  }
+}
+
+/**
+ * A string containing an HTTP URL not in a private domain.
+ */
+class HttpStringLiteral extends StringLiteral {
+  HttpStringLiteral() {
+    exists(string s | this.getValue() = s |
+      s = "http"
+      or
+      s.matches("http://%") and
+      not s.substring(7, s.length()) instanceof PrivateHostName and
+      not TaintTracking::localExprTaint(any(StringLiteral p |
+          p.getValue() instanceof PrivateHostName
+        ), this.getParent*())
+    )
+  }
+}
+
+/**
+ * Taint tracking configuration for HTTP connections.
+ */
+class HttpStringToUrlOpenConfig extends TaintTracking::Configuration {
+  HttpStringToUrlOpenConfig() { this = "HttpStringToUrlOpenConfig" }
+
+  override predicate isSource(DataFlow::Node src) { src.asExpr() instanceof HttpStringLiteral }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(FunctionCall fc |
+      fc.getTarget().getName() = ["system", "gethostbyname"] and
+      sink.asExpr() = fc.getArgument(0)
+      or
+      fc.getTarget().getName() = ["send", "URLDownloadToFile"] and
```

Comment:
`send` is a very common function name. Do we not have a model for the particular `send` you're trying to target (or is it a private model that we can't reach from here)?

Reply:
It should be `send` in the global namespace, i.e. https://man7.org/linux/man-pages/man2/send.2.html .  For now I've switched `getName` for `hasGlobalOrStdName` but I'll look into upgrading to using a proper model.

Reply:
There is a model for `send` and similar functions, but it's private.  I think I should just add this task to the ticket about private models ... except ... I can't find such a ticket.  Do you remember if we actually made it or just committed to doing it elsewhere?  I may have to create a ticket as the problem comes up quite frequently.

Reply:
I think we committed to doing it ... somewhere ... but I also can't find any mention of it 🤔 Feel free to create a dedicated issue for it.

Reply:
Created a ticket and called out `send` here: https://github.com/github/codeql-c-team/issues/737